FLT/Deformations/IsResidueAlgebra.lean

```diff
@@ -0,0 +1,109 @@
+import Mathlib.RingTheory.LocalRing.ResidueField.Basic
+import Mathlib.Logic.Function.Defs
+import FLT.Mathlib.RingTheory.LocalRing.Defs
+
+open Function
+open scoped TensorProduct
+
+namespace Deformation
+
+variable (𝓞 : Type*)
+  [CommRing 𝓞] [IsLocalRing 𝓞]
+
+local notation3:max "𝓴" 𝓞 => (IsLocalRing.ResidueField 𝓞)
+
+variable (A : Type*) [CommRing A] [Algebra 𝓞 A] [IsLocalRing A] [IsLocalHom (algebraMap 𝓞 A)]
+
+lemma injective : Injective (algebraMap (𝓴 𝓞) (𝓴 A)) := RingHom.injective (algebraMap (𝓴 𝓞) (𝓴 A))
+
+class IsResidueAlgebra : Prop where
+  isSurjective : Surjective (algebraMap 𝓞 (𝓴 A))
+
+namespace IsResidueAlgebra
+
+variable [IsResidueAlgebra 𝓞 A]
+
+lemma surjective : Surjective (algebraMap (𝓴 𝓞) (𝓴 A)) := by
+  have hsurj1 := (IsLocalRing.residue_surjective (R := 𝓞))
+  have hsurj2 := IsResidueAlgebra.isSurjective (𝓞 := 𝓞) (A := A)
+  exact (Function.Surjective.of_comp_iff (algebraMap (𝓴 𝓞) (𝓴 A)) hsurj1).mp hsurj2
+
+noncomputable def ringEquiv : (𝓴 𝓞) ≃+* (𝓴 A) := RingEquiv.ofBijective
+  (algebraMap (𝓴 𝓞) (𝓴 A)) ⟨injective 𝓞 A, surjective 𝓞 A⟩
+
+section Quotient
+
+variable {A} in
+omit [IsLocalRing A] in
+lemma Ideal.neq_top_of_nontrivial_quotient (I : Ideal A) [nontrivial : Nontrivial (A ⧸ I)] : I ≠ ⊤ := by
+  by_contra hc
+  have hsubsing := Ideal.Quotient.subsingleton_iff.mpr hc
+  exact false_of_nontrivial_of_subsingleton (A ⧸ I)
+
+variable {A} in
+def residue_of_quot (I : Ideal A) [nontrivial : Nontrivial (A ⧸ I)] : (𝓴 A) →+* 𝓴 (A ⧸ I) :=
+  Ideal.quotientMap (IsLocalRing.maximalIdeal (A ⧸ I)) (Ideal.Quotient.mk I) (by
+    rw [← Ideal.map_le_iff_le_comap]
+    suffices h : Ideal.map (Ideal.Quotient.mk I) (IsLocalRing.maximalIdeal A) ≠ ⊤ by
+      exact IsLocalRing.le_maximalIdeal h
+    rw [Ideal.ne_top_iff_one]
+    by_contra hcontra
+    have h := (Ideal.mem_map_iff_of_surjective (Ideal.Quotient.mk I) (Ideal.Quotient.mk_surjective)).mp hcontra
+    let x := h.choose
```

Comment:
Just do `cases h` (or, better, `obtain`), you don't need all this `choose` and `choose_spec` stuff, you're not making a definition.